[Console] Add Cursor class to control the cursor in the terminal

pierredup opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | TBD

Add a new `Cursor` class to the Console component, that allows to manipulate the cursor in the terminal.

Options to move the cursor include: moveUp, moveDown, moveLeft, moveRight, show, hide, save and restore position, get current position etc.

Example usage:

```php
$cursor = new Cursor($output);

$parts = [
    ['x' => 9, 'y' => 5, 'string' => '#'],
    ['x' => 9, 'y' => 7, 'string' => '#'],
    ['x' => 2, 'y' => 7, 'string' => '#####'],
    ['x' => 9, 'y' => 1, 'string' => '#######'],
    ['x' => 7, 'y' => 6, 'string' => '#'],
    ['x' => 9, 'y' => 2, 'string' => '#'],
    ['x' => 9, 'y' => 6, 'string' => '#'],
    ['x' => 9, 'y' => 3, 'string' => '#'],
    ['x' => 2, 'y' => 1, 'string' => '#####'],
    ['x' => 7, 'y' => 2, 'string' => '#'],
    ['x' => 9, 'y' => 4, 'string' => '#####'],
    ['x' => 1, 'y' => 2, 'string' => '#'],
    ['x' => 1, 'y' => 6, 'string' => '#'],
    ['x' => 2, 'y' => 4, 'string' => '#####'],
    ['x' => 1, 'y' => 3, 'string' => '#'],
    ['x' => 7, 'y' => 5, 'string' => '#'],
];

foreach ($parts as $part) {
    $cursor->moveToPosition($part['x'], $part['y']);
    $output->write($part['string']);

    usleep(200000);
}

$cursor->moveToPosition(1, 8);
```

Result:

![kapture 2018-05-31 at 12 39 10](https://user-images.githubusercontent.com/144858/40777952-abb7c252-64cf-11e8-80a7-7417676c8b0b.gif)

pierredup committed:
[Console] Add Cursor class to control the cursor in the terminal

javiereguiluz commented:
For reference purposes, I've look for the APIs for this feature in other projects:

1) Go libraries and .Net libraries have just a few methods:

https://godoc.org/github.com/nsf/termbox-go

```go
func HideCursor()
func SetCursor(x, y int)
```

https://godoc.org/github.com/gdamore/tcell

```go
func HideCursor()
func ShowCursor(x int, y int)
```

.Net (https://msdn.microsoft.com/en-us/library/system.console_methods(v=vs.110).aspx)

```csharp
SetCursorPosition(Int32, Int32)
```

And some properties:

```csharp
CursorLeft
CursorTop
CursorVisible
...
```

2) On the other hand, JS libraries define a ton of methods:

https://github.com/cronvel/terminal-kit/blob/master/doc/low-level.md#ref.movingCursor

```js
.saveCursor()
.restoreCursor()
.up(int n)
.down(int n)
.right(int n)
.left(int n)
.nextLine(int n)
.previousLine(int n)
.column(int x)
.scrollUp(int n)
.scrollDown(int n)
.scrollingRegion(top, bottom)
.resetScrollingRegion()
.moveTo(int x, int y)
.move(int x, int y)
.hideCursor()
.tabSet()
.tabClear()
.tabClearAll()
.forwardTab(int n)
.backwardTab(int n)
```

-----

So we need to think if we want a simple (but enough) API or a complex (but useful) API.

ogizanagi commented:
Also see the HOA Project Console `Cursor`: https://github.com/hoaproject/Console#cursor

yceruto commented on src/Symfony/Component/Console/Cursor.php:
```diff
@@ -0,0 +1,133 @@
+class Cursor
+{
+    /**
+     * @var OutputInterface
```
This can be removed since its type is discovered in the constructor.

yceruto commented on src/Symfony/Component/Console/Tests/CursorTest.php:
```diff
@@ -0,0 +1,215 @@
+     *
+     * @throws \Symfony\Component\Console\Exception\InvalidArgumentException
+     */
+    protected function getOutputStream($decorated = false)
```
`getOutputStream(bool $decorated = false): StreamOutput`?

yceruto reviewed (commented):
Thanks for this feature! I left two minor comments.

nicolas-grekas commented:
@pierredup see test failures :) Also, any comment when comparing the proposed API with the linked ones? Do any of them provide any idea to borrow?

chalasr reviewed (commented):
Can we update `QuestionHelper` to make it use this for autocompletion?
I'm fine with the current naming.

pierredup commented:
> Can we update QuestionHelper to make it use this for autocompletion?

@chalasr Done

chalasr commented on src/Symfony/Component/Console/Cursor.php:
```diff
@@ -0,0 +1,137 @@
+    public function clearScreen()
+    {
+        $this->output->write("\x1b[2J", false);
+    }
```
Does it really belong to the cursor?

chalasr reviewed (commented):
Can you add a CHANGELOG entry and rebase to see tests green?

fabpot commented on src/Symfony/Component/Console/Helper/QuestionHelper.php:
```diff
@@ -232,7 +235,7 @@ private function autocomplete(OutputInterface $output, Question $question, $inpu
                     --$i;
                     $fullChoice = self::substr($fullChoice, 0, $i);
                     // Move cursor backwards
```
The comment can be removed as the code is now explicit :)

fabpot commented on src/Symfony/Component/Console/Helper/QuestionHelper.php:
```diff
@@ -319,16 +322,16 @@ function ($match) use ($ret) {
+            $cursor->clearLine(true);
 
             if ($numMatches > 0 && -1 !== $ofs) {
                 // Save cursor position
```
same here

fabpot commented on src/Symfony/Component/Console/Helper/QuestionHelper.php:
```diff
@@ -319,16 +322,16 @@ function ($match) use ($ret) {
                 // Write highlighted text, complete the partially entered response
                 $charactersEntered = \strlen(trim($this->mostRecentlyEnteredValue($fullChoice)));
                 $output->write('<hl>'.OutputFormatter::escapeTrailingBackslash(substr($matches[$ofs], $charactersEntered)).'</hl>');
                 // Restore cursor position
```
same :)

fabpot reviewed (changes requested):
LGTM.

Can we also have methods to get the max x and max y of the console/cursor? I think we already have the code in Symfony, that would be good to have it here as well.

Maybe we can simplify the progress bar code as well.

pierredup commented:
I've rebased from master, and updated the Progress bar to use the Cursor class as well.

> Can we also have methods to get the max x and max y of the console

This is currently available in the [Terminal](https://github.com/symfony/symfony/blob/ae35526/src/Symfony/Component/Console/Terminal.php) class for this component, so I'm not sure if it should be in the Cursor class as well

chalasr commented:
Agree about keeping the terminal size logic in the `Terminal` class.
@pierredup Some related tests are failing, can you have a look?

chalasr commented on src/Symfony/Component/Console/Cursor.php:
```diff
@@ -0,0 +1,137 @@
+        $sttyMode = shell_exec('stty -g');
+        shell_exec('stty -icanon -echo');
+
+        fwrite($this->input, "\033[6n");
```
`@fwrite`  
see https://github.com/symfony/symfony/pull/36222 and https://www.php.net/manual/en/migration74.incompatible.php#migration74.incompatible.core.fread-fwrite`

pierredup replied:
Thanks @chalasr

chalasr reviewed (approved):
LGTM :) Missing CHANGELOG entry

fabpot commented:
Thanks @pierredup for working on this feature, this is much appreciated.
